terra/contracts/token-bridge/src/contract.rs

```diff
@@ -121,8 +121,41 @@ const CHAIN_ID: u16 = 3;
 const WRAPPED_ASSET_UPDATING: &str = "updating";
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
-    Ok(Response::new())
+pub fn migrate(deps: DepsMut, env: Env, _msg: MigrateMsg) -> StdResult<Response> {
+    let messages = migrate_wrapped_assets(deps, env, 767)?;
+    let count = messages.len();
+
+    Ok(Response::new()
+        .add_messages(messages)
+        .add_attribute("migrate", "upgrade cw20 wrappers")
+        .add_attribute("count", count.to_string()))
+}
+
+/// Migrate all wrapped assets to a new code id.
+/// This function should be called in [`migrate`].
+fn migrate_wrapped_assets(deps: DepsMut, _env: Env, new_code_id: u64) -> StdResult<Vec<CosmosMsg>> {
```

Comment:
Can we add some tests for this?  It looks like we can set up a unit test environment with [`cosmwasm_std::testing::mock_dependencies`](https://docs.rs/cosmwasm-std/0.16.6/cosmwasm_std/testing/fn.mock_dependencies.html) and [`cosmwasm_std::testing::mock_env`](https://docs.rs/cosmwasm-std/0.16.6/cosmwasm_std/testing/fn.mock_env.html).

In addition to the normal case, what should be the behavior if `wrapped_asset_code_id == new_code_id`?  What happens if we deploy a new version of the contract without removing this migration code?  Or if we accidentally deploy an older version of the contract?  It feels like we need some basic check to make sure that we only run the migration when going from version X -> version Y and either skip or throw an error in any other case.

Is there a way for us to prevent this from happening again?  For example, we could define the code id as a constant somewhere and add a check in `initialize` that will error out if the stored code id doesn't match the constant (and then it's easy to add a test for that behavior).

Reply:
> Is there a way for us to prevent this from happening again?

Here's the code that was ran during the previous migration: https://github.com/certusone/wormhole/commit/c832b123fcfb017d55086cb4d71241370ed270c6#diff-4d1ad320d9f5ab48ff71ac9974eb9a9b10739c15e740a137d85f4fd0354e98f5R124-R144
this current PR is supposed to prevent this issue from happening again by setting the state too in `migrate_wrapped_assets`.

> In addition to the normal case, what should be the behavior if wrapped_asset_code_id == new_code_id?

We could do a check to reject that moving forward. If we forget, that's a no-op, because upgrading the wrapped asset contract to a given id is idempotent (since the wrapped asset contracts have an empty `migrate` handler:
https://github.com/certusone/wormhole/blob/ddd2a60ab8da54966f5383df88eaf5c5bbe73973/terra/contracts/cw20-wrapped/src/contract.rs#L223-L226)

However, the migrate code of the token bridge in general is not idempotent (since it can run arbitrary code), so we must make sure that once it's done, the code is removed. Added a note about this too.

> Or if we accidentally deploy an older version of the contract?

This would require putting the wrong code id into the VAA. When the guardians vote on the VAA, they independently verify that the checksum of the code that's being deployed (by building it locally) matches the checksum of the binary that's stored on-chain at the given code id. There's of course a human element here, but I think the real risk is upgrading to a contract that doesn't belong to the token bridge. I don't know of a good way to protect against that, but it would be worthwhile to investigate.

> Can we add some tests for this? It looks like we can set up a unit test environment with [cosmwasm_std::testing::mock_dependencies](https://docs.rs/cosmwasm-std/0.16.6/cosmwasm_std/testing/fn.mock_dependencies.html) and [cosmwasm_std::testing::mock_env](https://docs.rs/cosmwasm-std/0.16.6/cosmwasm_std/testing/fn.mock_env.html).

As part of this PR, we should investigate that testing framework.

Reply:
I've fixed all the terra tests in #1046 so it should be possible to add tests now (after applying that PR locally).  I think the tests in cw20-wrapped/{src/contract.rs,tests/integration.rs} provide a good template for testing the contract functions.